content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,320 @@
+    margin: 10px 0px;
+    cursor: pointer;
+    box-shadow: 0px 2px 2px lightgray;
+    transition: ease background-color 250ms;
+    &:hover {
+      background-color: ${(props) => theme[props.theme].hover};
+    }
+    &:disabled {
+      cursor: default;
+      opacity: 0.7;
+    }
+  `;
+
+const clickMe = () => {
+  alert("A new shiny feature pops up!");
+};
+
+class App extends Component {
+  constructor() {
+    super()
+    this.state = {}
+  }
+  componentDidMount() {
+    const user = {
+      // UI based logged-in user
+      key: 'john_doe'
+    }
+    // SDK requires Client-side ID for UI call
+    this.ldclient = LDClient.initialize('62e*********************', user);
+    this.ldclient.on('ready', this.onLaunchDarklyUpdated.bind(this));
+    this.ldclient.on('change', this.onLaunchDarklyUpdated.bind(this));
+  }
+  onLaunchDarklyUpdated() {
+    this.setState({
+      featureFlags: {
+        showShinyNewFeature: this.ldclient.variation('show-shiny-new-feature', false)
+      }
+    })
+  }
+  render() {
+    if (!this.state.featureFlags) {
+      return <div className="App">Loading....</div>
+    }
+    return (
+      <div className="App">
+        <div>{ 
+            this.state.featureFlags.hideUser
+              ? <Button theme='blue' onClick={clickMe}>Shiny New Feature</Button>
+              : '' }
+        </div>
+    );
+  }
+}
+
+export default App;
+```
+
+To execute code only when the LaunchDarkly client is ready, we use the `ldclient.on()` function.
+
+With `this.ldclient.on('ready', ...)`, we subscribe to the `ready` event which will fire once the LaunchDarkly client has received the state of all feature flags from the server.
+
+With `this.ldclient.on('change', ...)`, we subscribe to the `change` event which will fire once the LaunchDarkly client has received the state of all feature flags that has got updated or changed in the LaunchDarkly UI.
+
+Next we can run the following command to start the UI:
+
+```bash
+npm start
+```
+
+When the user is not set in the LaunchDarkly UI, our React UI looks like below:
+
+{{% image alt="React User List" src="images/posts/nodejs-frontend-launchdarkly/Button.png" %}}
+
+Now based upon the user information, this flag will return true if the user is “John Doe”. It will return false for any other user who would try to access this UI. In a real production implementation, this name will be fetched from the logged-in session of the user and matched in Launchdarkly with its ID or username.
+
+## Sorting Data using Feature Flags
+
+Next we will try to sort our content in the UI using our feature flags. 
+
+Let's imagine we have a list of users, each with a timestamp. We can sort the users naturally or based upon that timestamp. We will create a feature flag variation in LaunchDarkly UI with a string type and try to use it in our code.
+
+{{% image alt="Boolean Feature Flag LaunchDarkly" src="images/posts/nodejs-frontend-launchdarkly/User_List_Sorting.png" %}}
```

Comment:
This screenshot also still shows the old feature flag before your changes. Can you update it?

Reply:
Updated the screenshot.